pkg/extensions/sync/remote.go

```diff
@@ -5,192 +5,364 @@ package sync
+	regclient *regclient.RegClient,
+) (
+	ispec.Descriptor, error,
+) {
+	var index ispec.Index
 
-	digest, ok := parseReference(reference)
-	if ok {
-		namedRepoRef, err = dockerReference.WithDigest(repoRef, digest)
+	index.SchemaVersion = 2
+	index.Manifests = []ispec.Descriptor{}
+	index.MediaType = ispec.MediaTypeImageIndex
+
+	indexer, ok := man.(manifest.Indexer)
+	if !ok {
+		return ispec.Descriptor{}, zerr.ErrMediaTypeNotSupported
+	}
+
+	ociIndex, err := manifest.OCIIndexFromAny(man.GetOrig())
+	if err != nil {
+		return ispec.Descriptor{}, zerr.ErrMediaTypeNotSupported
+	}
+
+	manifests, err := indexer.GetManifestList()
+	if err != nil {
+		return ispec.Descriptor{}, zerr.ErrMediaTypeNotSupported
+	}
+
+	for _, manDesc := range manifests {
+		ref := imageReference
+		ref.Digest = manDesc.Digest.String()
+
+		manEntry, err := regclient.ManifestGet(ctx, ref)
 		if err != nil {
-			return nil, err
+			return ispec.Descriptor{}, err
 		}
-	} else {
-		namedRepoRef, err = dockerReference.WithTag(repoRef, reference)
-		if err != nil {
-			return nil, err
+
+		regclient.Close(ctx, manEntry.GetRef())
+
+		var desc ispec.Descriptor
+
+		switch manEntry.GetDescriptor().MediaType {
+		case mediatype.Docker2Manifest:
+			desc, err = convertDockerManifestToOCI(ctx, manEntry, manDesc, ref, regclient)
+			if err != nil {
+				return ispec.Descriptor{}, err
+			}
+
+		case mediatype.Docker2ManifestList:
+			desc, err = convertDockerListToOCI(ctx, manEntry, ref, regclient)
+			if err != nil {
+				return ispec.Descriptor{}, err
+			}
+		default:
+			return ispec.Descriptor{}, err
 		}
+
+		index.Manifests = append(index.Manifests, desc)
 	}
 
-	imageRef, err := docker.NewReference(namedRepoRef)
-	if err != nil {
-		registry.log.Err(err).Str("transport", docker.Transport.Name()).Str("reference", namedRepoRef.String()).
-			Msg("cannot obtain a valid image reference for given transport and reference")
+	index.Annotations = ociIndex.Annotations
 
-		return nil, err
+	indexBuf, err := json.Marshal(index)
+	if err != nil {
+		return ispec.Descriptor{}, err
 	}
 
-	return imageRef, nil
+	indexDesc := toOCIDescriptor(man.GetDescriptor())
+
+	indexDesc.MediaType = ispec.MediaTypeImageIndex
+	indexDesc.Digest = godigest.FromBytes(indexBuf)
+	indexDesc.Size = int64(len(indexBuf))
+
+	return indexDesc, nil
 }
 
-func (registry *RemoteRegistry) GetManifestContent(imageReference types.ImageReference) (
-	[]byte, string, digest.Digest, error,
+func convertDockerManifestToOCI(ctx context.Context, man manifest.Manifest, desc descriptor.Descriptor,
+	imageReference ref.Ref, regclient *regclient.RegClient,
+) (
+	ispec.Descriptor, error,
 ) {
-	imageSource, err := imageReference.NewImageSource(context.Background(), registry.GetContext())
+	imager, ok := man.(manifest.Imager)
+	if !ok {
+		return ispec.Descriptor{}, zerr.ErrMediaTypeNotSupported
+	}
+
+	var ociManifest ispec.Manifest
```

Comment:
Don't we want to also return this object? Same for the index object which is not returned by the other function.